### PR TITLE
Adds side-by-side diff for images

### PR DESCRIPTION
**Pull request by saitho**
This adds a basic image diff and prints the width of the images (before and after). This is part of what was requested in #6255 .

Screenshot:
image-diff

The side-by-side view for images will be active regardless whether split view is active or not.

### CODE REVIEW

**Comment by lafriks** [comment]
Probably better would be to have `Width: 1156, Height: 412` instead of `W: 1156, H: 412`

Also what will happen with very large images?

**Comment by saitho** [comment]
The images are embedded how they were uploaded in the repository, i.e. in their full dimensions.
The width is limited by `width: 100%` in `.diff-file-box.file-content img` which is already present.

File: modules/templates/helper.go
Comment (lafriks) [inline modules/templates/helper.go]: I don't think this is correct way to get imagine info, it should be taken from server git repo

Comment (saitho) [reply modules/templates/helper.go]: Agreed.

File: routers/repo/pull.go
Comment (saitho) [inline routers/repo/pull.go:570]: Note that I changed "headTarget" to "baseTarget", as the "before" paths should point to the forked repository.

File: modules/git/commit.go
Comment (jonasfranz) [inline modules/git/commit.go]: please return error instead of nil

**Review by jonasfranz — CHANGES_REQUESTED** [review]

**Comment by silverwind** [comment]
If it's not too hard, I'd request adding a size below each image. Would be useful to compare size optimizations:

Screenshot 2019-05-02 at 09 10 06

**Comment by saitho** [comment]
> If it's not too hard, I'd request adding a size below each image. Would be useful to compare size optimizations

I'll look into that. :) I assume that the filesystem library is able to do that.

**Comment by saitho** [comment]
image-diff-filesize

There we go. :)

File: modules/git/commit.go
Comment (lafriks) [inline modules/git/commit.go]: do not update copyright year

File: modules/git/commit.go
Comment (lafriks) [inline modules/git/commit.go:189]: Reader should be closed imho

```suggestion
	defer reader.Close()
	config, _, err := image.DecodeConfig(reader)
```

File: routers/repo/pull.go
Comment (lafriks) [inline routers/repo/pull.go]: same here

**Comment by saitho** [comment]
@lafriks Done :)

File: routers/repo/commit.go
Comment (lafriks) [inline routers/repo/commit.go]: You have already set it few lines before

**Review by lafriks — APPROVED** [review]

File: modules/git/commit.go
Comment (lunny) [inline modules/git/commit.go]: this line should be in line 188, because if line 190 returned, the `reader` will not be `Close`.

File: routers/repo/commit.go
Comment (lunny) [inline routers/repo/commit.go:255]: Maybe `ctx.ServerError(` is better.

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by jonasfranz** [comment]
conflicted

**Comment by silverwind** [comment]
You CSS for some reason does not include autoprefixer rules. Can you rebase against master again and do

````
rm -rf node_modules
npm install
make css
````

**Comment by silverwind** [comment]
Thanks, that fixed it.

**Comment by lunny** [comment]
conflicted

**Review by lunny — APPROVED** [review]

**Review by appleboy — APPROVED** [review]

**Comment by lunny** [comment]
need @jonasfranz 's approval

**Review by sapk — APPROVED** [review]

**Comment by jonasfranz** [comment]
What happens if I commit big images? We should maybe not open those if they're too big.

**Comment by saitho** [comment]
> What happens if I commit big images? We should maybe not open those if they're too big.

Big as in file size?
Not sure why file size would be an issue. I guess travelling programmers which rely on cellular network with limited bandwidth or traffic volume would welcome not loading a 5 MB image. Is this a use case Gitea generally considers?

Big as in large dimensions? See:
> The images are embedded how they were uploaded in the repository, i.e. in their full dimensions.
> The width is limited by `width: 100%` in `.diff-file-box.file-content img` which is already present.

**Comment by silverwind** [comment]
I don't see the need for a file size limit. It's the responsibility of the developer to not commit overly large files imho. There is no inherent performance issue for big images as there is for big diffs.

**Comment by kolaente** [comment]
@jonasfranz please re-review

File: options/locale/locale_en-US.ini
Comment (silverwind) [inline options/locale/locale_en-US.ini]: Minor nit, but I'd prefer just `Size` over `File Size`

**Comment by lunny** [comment]
@saitho Please resolve the conflicts.
@jonasfranz Since they are two image medias, they will be loaded asynchrony by web browser. They will not block the main page loading when review.

**Comment by jonasfranz** [comment]
1. conflicted
2. When I create a new empty repository and commit an image in the first commit, there will  be a corrupted old image. (see screenshot and code)
Screenshot_2019-07-09 geissler
```html
<td class="halfwidth center">
 		<a href="%25!s%28%3cnil%3e%29/gei%c3%9fler.jpeg" target="_blank">
 			<img src="%25!s%28%3cnil%3e%29/gei%c3%9fler.jpeg" class="border red">
 		</a>
 	</td>
```

File: routers/repo/pull.go
Comment (jonasfranz) [inline routers/repo/pull.go]: Missing check? What happens if there is no startCommitID because it's the first commit? (see my last comment)

File: routers/repo/pull.go
Comment (jonasfranz) [inline routers/repo/pull.go:571]: Missing check? What happens if there is no startCommitID because it's the first commit? (see my last comment)

**Review by jonasfranz — DISMISSED** [review]

**Comment by saitho** [comment]
Okay. I'll do that on the weekend, as the PR isn't relevant for 1.9.0. ;)

Jonas Franz <notifications@github.com> schrieb am Di., 9. Juli 2019, 22:21:

**Comment by saitho** [comment]
Hey there. Sorry for the long delay.
Wasn't really motivated to work on Gitea - now I am. :)

>     2. When I create a new empty repository and commit an image in the first commit, there will  be a corrupted old image. (see screenshot and code)
>        Screenshot_2019-07-09 geissler

This happens when the image is newly added to the repository (i.e. is not in the index of the base commit). Fixed by only displaying the image if it exists in the base commit index.

File: routers/repo/pull.go
Comment (saitho) [reply routers/repo/pull.go:571]: If there was only one commit you wouldn't be able to compare it at all. ;)
What you experienced is the scenario when the image is not available in the previous commit index.
Fixed.

**Comment by saitho** [comment]
Ready for review @jonasfranz

**Review by lafriks — APPROVED** [review]

**Comment by guillep2k** [comment]
Could this PR have latest master merged in and CI restarted? Looks like CI problems are not from this PR.

**Comment by lunny** [comment]
Since @jonasfranz didn't response this one for a month and there are two approvals on this, I think we can merge this one when CI PASS.

**Comment by guillep2k** [comment]
@lunny @lafriks It looks like the tests I've added to the UTF functions are quite problematic. They fail only _from time to time_. I guess it's a LOCALE setting. Can a specific LOCALE be forced in the Makefile for the tests? Something like `LOCALE=en_US.UTF-8` or any other, as long as it is always the same.

**Comment by guillep2k** [comment]
Otherwise, I'd remove most of the UTF8 tests. I don't think they're that much important.

**Comment by lunny** [comment]
make L-G-T-M work

**Pull request merged**
